### PR TITLE
Patched stringToBytes32 which must pad on the left

### PR DESCRIPTION
**Pull request by james-toussaint**

### CODE REVIEW

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 1]
```diff
@@ -98,23 +98,25 @@ public static boolean isNonZeroedBytes32(String hexString) {
...
+     *
+     * @param hexString hexadecimal string
+     * @return bytes32
+     */
     public static byte[] stringToBytes32(String hexString) {
-        byte[] stringBytes = stringToBytes(hexString);
-        if (isBytes32(stringBytes)) return stringBytes;
-
-        String cleanString = Numeric.cleanHexPrefix(hexString);
-        String padded = padRight(cleanString, 64 - cleanString.length());
```
Comment (james-toussaint) [inline src/main/java/com/iexec/common/utils/BytesUtils.java:111]: Value must be padded on the left

File: src/test/java/com/iexec/common/utils/BytesUtilsTests.java [hunk 2]
```diff
@@ -159,21 +159,29 @@ public void shouldNotBeAByte32Null() {
...
-        assertEquals(bytesToString(bytes32), hexaString);
+        assertEquals(32, bytes32.length);
+        assertArrayEquals(this.bytes, bytes32);
+        assertEquals(hexaString, bytesToString(bytes32));
     }
 
     @Test
-    public void shouldPadStringToBeBytes32() {
-        String notBytes32String = "0xabc";
-        String bytes32String = "0xabc0000000000000000000000000000000000000000000000000000000000000";
```
Comment (james-toussaint) [inline src/test/java/com/iexec/common/utils/BytesUtilsTests.java:172]: Awkward test

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 3]
```diff
@@ -98,23 +98,25 @@ public static boolean isNonZeroedBytes32(String hexString) {
...
-        return Numeric.hexStringToByteArray(padded);
-    }
-
-    public static String padRight(String s, int n) {
-        if (n <= 0) return s;
-        String zeros = new String(new char[n]).replace('\0', '0');
-        return s + zeros;
+        if (!isHexStringWithPrefix(hexString)) {
+            return new byte[0];
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]: I think it's more appropriate to throw an `IllegalArgumentException` here.

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 4]
```diff
@@ -98,23 +98,25 @@ public static boolean isNonZeroedBytes32(String hexString) {
...
+        }
+        if (isBytes32(hexString)) {
+            return Numeric.hexStringToByteArray(hexString);
+        }
+        return Numeric.hexStringToByteArray(Numeric.toHexStringWithPrefixZeroPadded(
+                Numeric.toBigInt(hexString), 64));
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]: Will the string be shrunk if its size is greater than 64 ?

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 5]
```diff
@@ -98,23 +98,25 @@ public static boolean isNonZeroedBytes32(String hexString) {
...
 
-    // this adds zeros to the left of the hex string to make it bytes32
+    /**
+     * Convert any hex string input into a bytes32 (full 32 length).
+     * Eventually pad with zeros on the left to ensure 32 length.
+     *
+     * @param hexString hexadecimal string
+     * @return bytes32
+     */
     public static byte[] stringToBytes32(String hexString) {
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]:
```suggestion
    public static byte[] hexStringToBytes32(String hexString) {
```

File: src/main/java/com/iexec/common/utils/BytesUtils.java (hunk 3 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: I agree!

File: src/main/java/com/iexec/common/utils/BytesUtils.java (hunk 4 above)
Comment (james-toussaint) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: We now throw for it

File: src/main/java/com/iexec/common/utils/BytesUtils.java [hunk 6]
```diff
@@ -102,16 +104,24 @@ public static boolean isBytes32(byte[] bytes) {
...
      */
-    public static byte[] stringToBytes32(String hexString) {
-        if (!isHexStringWithPrefix(hexString)) {
-            return new byte[0];
+    public static byte[] hexStringToBytes32(String hexString) {
+        if (!isHexStringWithPrefix(hexString)
+                || hexString.length() > BYTES_32_HEX_STRING_WITH_PREFIX_SIZE) {
+            throw new IllegalArgumentException(String.format("Input string " +
+                    "should be an hexadecimal string with 0x prefix and 66 " +
+                    "length [input:%s]", hexString));
```
Comment (zguesmi) [inline src/main/java/com/iexec/common/utils/BytesUtils.java]:
```suggestion
                    "as maximum length [input:%s]", hexString));
```

**Review by zguesmi — APPROVED** [review]
Thanks!

File: src/main/java/com/iexec/common/utils/BytesUtils.java (hunk 4 above)
Comment (zguesmi) [reply src/main/java/com/iexec/common/utils/BytesUtils.java]: Thanks

**Pull request merged**